### PR TITLE
Restructure spectrum package and command line tool

### PR DESCRIPTION
**Pull request by joleroi**
this PR changes

the structure of gammapy.spectrum 
* SpectrumObservation is now a pure container class with classmethods doing the work
* SpectrumAnalysis -> SpectrumExtraction. Only algorithms
* Add SpectrumObservationList to represent a list of observation. Dummy methods for stacking
* Fitting is completely independent now and lives in spectrum.spectrum_fit
* Add utility function to spectrum.result for overplotting and comparison tables

the structire of the gammapy-spectrum command line tool
* Add subcommands extract fit display plot

### CODE REVIEW

**Comment by cdeil** [comment]
@joleroi - Now that you've made a change in gammapy-extra, but this PR from gammapy hasn't landed yet, the build is broken:
https://travis-ci.org/gammapy/gammapy/jobs/107459817#L885

I really need to change the setup, so that gammapy-extra is a git submodule ... this will give us versioning for gammapy-extra and sanity.

**Comment by joleroi** [comment]
@cdeil Sorry for breaking everything :grin: 

Can you please have a look at gammapy/spectrum/results.py (and for now only at this). There are not test yet (only one example) but next I want to add a test that runs an entire crab analysis and does stuff with the results, so the results classes will be tested in this high-level test (You said: High-level tests over unit-tests).

Is there any functionality in spectrum/spectrum_results.py that you need and is not covered in the new stuff? Otherwise I would think about removing it, because having spectrum_results.py and results.py is confusing (Luigi already stumbled over this).

**Comment by joleroi** [comment]
You only want to do the tutorial once this PR is merged, right?

**Comment by cdeil** [comment]
> Can you please have a look at gammapy/spectrum/results.py (and for now only at this). 

I had a quick look. It's super-hard to review this in isolation and give useful feedback.

I'd much prefer to merge in master, and let me just use this for an hour tomorrow and look around the docs and code. Then I can give suggestions like I did in #419.

@joleroi - Would this be OK for you?

> You said: High-level tests over unit-tests

Yes, I don't think unit tests make sense as long as the API is still heavily changing. That's not just me ... you can find many opinions that unit tests don't make sense during prototyping, because you'd spend so much time refactoring the tests over and over as the API changes.

> Is there any functionality in spectrum/spectrum_results.py that you need and is not covered in the new stuff?

I'm not sure ... could you leave it in for now and I'll have a detailed look tomorrow?

File: gammapy/spectrum/tests/test_spectrum_fit.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    pha_list = [pha1, pha2]
+    fit = SpectrumFit(pha_list)
+    fit.model = 'PL'
+    fit.energy_threshold_low = '100 GeV'
+    fit.energy_threshold_high = '10 TeV'
+    fit.run(method='sherpa')
+    assert fit.result.spectral_model == 'PowerLaw'
+
+
+def test_spectral_fit_using_config(tmpdir):
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_spectrum_fit.py]: Add this here:
```
@requires_dependency('sherpa')
@requires_data('gammapy-extra')
```
https://travis-ci.org/gammapy/gammapy/jobs/107466758#L889

File: gammapy/spectrum/tests/test_spectrum_fit.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+from astropy.tests.helper import pytest
+from numpy.testing import assert_allclose
+
+from gammapy.spectrum import run_spectrum_extraction_using_config
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_spectrum_fit.py]: Use relative imports within Gammapy ... this needs to be changed in several files in this PR.

**Comment by cdeil** [comment]
+1,937 −1,409 lines -> ["looks fine"](https://alexgaynor.net/2015/dec/29/shrinking-code-review/) :-)

File: gammapy/spectrum/tests/test_spectrum_fit.py (hunk 2 above)
Comment (joleroi) [inline gammapy/spectrum/tests/test_spectrum_fit.py]: Stupid PyCharm ...

**Comment by cdeil** [comment]
What about this docs page?
https://gammapy.readthedocs.org/en/latest/spectrum/spectral_fitting.html

Has it been superceded by this one?
https://github.com/gammapy/gammapy/tree/master/docs/tutorials/gammapy-spectrum

Or do you want to keep both?

When I wanted to try out the new spectrum fitting code, https://gammapy.readthedocs.org/en/latest/spectrum/spectral_fitting.html is what I found first,
but it looks like it doesn't quite work any more:
```
$ gammapy-spectrum crab_config.yaml
Usage: gammapy-spectrum [OPTIONS] COMMAND [ARGS]...

Error: No such command "crab_config.yaml".
$ gammapy-spectrum 
Usage: gammapy-spectrum [OPTIONS] COMMAND [ARGS]...

  Gammapy tool for spectrum extraction and fitting.

  Examples
  --------
  
  TODO

Options:
  -h, --help  Show this message and exit.

Commands:
  all      Fit spectral model to 1D spectrum
  display  Display results table of one or several...
  extract  Extract 1D spectral information from event...
  fit      Fit spectral model to 1D spectrum
  pipe     Run spectrum analysis pipeline
  plot     Plot spectrum results file
$ gammapy-spectrum all crab_config.yaml
Traceback (most recent call last):
  File "/Users/deil/Library/Python/3.4/bin/gammapy-spectrum", line 9, in <module>
    load_entry_point('gammapy', 'console_scripts', 'gammapy-spectrum')()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/Users/deil/code/gammapy/gammapy/scripts/spectrum.py", line 67, in all_spectrum
    fit, analysis = run_spectrum_analysis_using_config(config)
  File "/Users/deil/code/gammapy/gammapy/spectrum/spectrum_pipe.py", line 113, in run_spectrum_analysis_using_config
    analysis = run_spectrum_extraction_using_config(config)
  File "/Users/deil/code/gammapy/gammapy/spectrum/spectrum_extraction.py", line 632, in run_spectrum_extraction_using_config
    config = config['extraction']
KeyError: 'extraction'
```

**Comment by joleroi** [comment]
No this is in fact outdated. I would like to keep it for "low-level" docs. I will remove the broken part in the next PR. Could you try the new tutorial?

**Comment by cdeil** [comment]
I'm trying the new tutorial now ... can you come by so that we discuss a bit? ... I think it's easier than via Github for this.

File: gammapy/spectrum/results.py [hunk 3]
```diff
@@ -503,31 +686,83 @@ def plot_spectrum(self, energy_unit='TeV', flux_unit='cm-2 s-1 TeV-1',
...
         for key in val:
-            temp = QTable(self[key].to_table())
-            new_row = list()
-            for col in table.columns:
-                new_row.append(temp[col].to(table[col].unit).value)
-            table.add_row(new_row)
-
+            l.append(self[key].to_table(**kwargs))
+        table = vstack(l, join_type='outer')
         table.add_column(analyses, index=0)
```
Comment (cdeil) [inline gammapy/spectrum/results.py]: With Python 3 and Astropy and Gammapy master I'm getting this error when I follow along the new tutorial:
```
$ gammapy-spectrum display crab_analysis/spectrum_stats.yaml spectrum_stats_23592.yaml
Traceback (most recent call last):
  File "/Users/deil/Library/Python/3.4/bin/gammapy-spectrum", line 9, in <module>
    load_entry_point('gammapy', 'console_scripts', 'gammapy-spectrum')()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/Users/deil/code/gammapy/gammapy/scripts/spectrum.py", line 99, in display_spectrum
    t = master.to_table(format=format)
  File "/Users/deil/code/gammapy/gammapy/spectrum/results.py", line 762, in to_table
    table.add_column(analyses, index=0)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/astropy/table/table.py", line 1387, in add_column
    self.add_columns([col], [index], rename_duplicate=rename_duplicate)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/astropy/table/table.py", line 1495, in add_columns
    self._init_from_cols(newcols)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/astropy/table/table.py", line 685, in _init_from_cols
    lengths = set(len(col) for col in cols)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/astropy/table/table.py", line 685, in <genexpr>
    lengths = set(len(col) for col in cols)
TypeError: len() of unsized object
```

**Comment by cdeil** [comment]
This is the template render package I mentioned:
http://jinja.pocoo.org/docs/dev/intro/#basic-api-usage

As I said, I'd start by adding the code to make a webpage with the most common plot and numbers for a single analysis, with the plots FitSpectrum generates and prints on the console.
And then, if you want it for your work, expand / improve the subcommands that compare different analyses in text output and plots.
Wanting to fit PL, ECPL and CPL spectral shapes is super common ... this is the one thing where exposing it at the "single analysis config file" level (e.g. as a list in the yaml file) could be useful, or at least it has to be easy / documented how to do this.

**Pull request merged**